Datastore revamp 7: garbage collection

teh-cmc opened the pull request:
This PR implements garbage collection in the new store and therefore puts an end to the revamp series.

- Implement new `RowId`-based GC
    - Focused on correctness, millions of opportunities to improve performance in the future
    - Full support for out of order data
    - Row-level granularity: not tied to the bucket size anymore
- Lots of improvements to size measurements
    - (Very) Important to reduce the mismatch between what the GC sees vs. the ground truth reported by the global allocator
- Kill a bunch of unrelated bugs that were found when stress-testing the GC

Highlighted a bunch of issues:
- #1803 
- #1804 
- #1805 
- #1807 


---

1. #1727 
2. #1735
3. #1739 
4. #1785 
5. #1791 
6. #1795
7. #1801

---

On top of #1795
Part of #1619 

Fixes #894

Future work:
- #1823 

---

- [x] Self-review
- [x] No bench regression
- [x] No test regression
- [x] TODO all the things that fit into #1619 
- [x] open issues for everything else

---

_(Fixed some bugs last minute that improved the graphs in the following demos, but you get the general idea)_

https://user-images.githubusercontent.com/2910679/230886287-f0dbc880-88bc-42bb-9656-a2276c3d61f8.mp4

https://user-images.githubusercontent.com/2910679/230885453-8ecec0ff-764d-487b-849c-848449df588c.mp4

https://user-images.githubusercontent.com/2910679/230887955-c6ced7d3-74c0-4cd8-9a28-5ca30c63a471.mp4

---

<details>
  <summary>Benchmarks</summary>


| Benchmark suite | Current: a889abf21982535675d0d2ff0cc991778ddba14c | Previous: e3f572f6e148780ebd75c5ce9d32f58c4b5da2c6 | Ratio |
|-|-|-|-|
| `datastore/num_rows=1000/num_instances=1000/packed=false/insert/default` | `2493172` ns/iter (`± 32525`) | `11575476` ns/iter (`± 688622`) | `0.22` |
| `datastore/num_rows=1000/num_instances=1000/packed=false/latest_at/default` | `370` ns/iter (`± 3`) | `1809` ns/iter (`± 15`) | `0.20` |
| `datastore/num_rows=1000/num_instances=1000/packed=false/latest_at_missing/primary/default` | `262` ns/iter (`± 0`) | `278` ns/iter (`± 0`) | `0.94` |
| `datastore/num_rows=1000/num_instances=1000/packed=false/latest_at_missing/secondaries/default` | `420` ns/iter (`± 0`) | `440` ns/iter (`± 2`) | `0.95` |
| `datastore/num_rows=1000/num_instances=1000/packed=false/range/default` | `2625131` ns/iter (`± 45606`) | `12352247` ns/iter (`± 772207`) | `0.21` |
| `mono_points_arrow/generate_message_bundles` | `27981679` ns/iter (`± 427787`) | `48276625` ns/iter (`± 1397029`) | `0.58` |
| `mono_points_arrow/generate_messages` | `114218442` ns/iter (`± 914602`) | `172700989` ns/iter (`± 1490257`) | `0.66` |
| `mono_points_arrow/encode_log_msg` | `155130549` ns/iter (`± 2102982`) | `200215428` ns/iter (`± 1409723`) | `0.77` |
| `mono_points_arrow/encode_total` | `290410867` ns/iter (`± 1482761`) | `423220801` ns/iter (`± 2426710`) | `0.69` |
| `mono_points_arrow/decode_log_msg` | `178057447` ns/iter (`± 673276`) | `253529959` ns/iter (`± 1198838`) | `0.70` |
| `mono_points_arrow/decode_message_bundles` | `60925200` ns/iter (`± 507883`) | `84116550` ns/iter (`± 1373716`) | `0.72` |
| `mono_points_arrow/decode_total` | `236696708` ns/iter (`± 1062047`) | `344891240` ns/iter (`± 2570130`) | `0.69` |
| `mono_points_arrow_batched/generate_message_bundles` | `21948129` ns/iter (`± 990466`) | `41535573` ns/iter (`± 1799262`) | `0.53` |
| `mono_points_arrow_batched/generate_messages` | `4580723` ns/iter (`± 159754`) | `8656223` ns/iter (`± 993434`) | `0.53` |
| `mono_points_arrow_batched/encode_log_msg` | `1379279` ns/iter (`± 4409`) | `1658024` ns/iter (`± 5637`) | `0.83` |
| `mono_points_arrow_batched/encode_total` | `29621541` ns/iter (`± 1163999`) | `53520656` ns/iter (`± 2841832`) | `0.55` |
| `mono_points_arrow_batched/decode_log_msg` | `784267` ns/iter (`± 2603`) | `945416` ns/iter (`± 3479`) | `0.83` |
| `mono_points_arrow_batched/decode_message_bundles` | `7717233` ns/iter (`± 186748`) | `12660668` ns/iter (`± 673928`) | `0.61` |
| `mono_points_arrow_batched/decode_total` | `9002045` ns/iter (`± 313841`) | `14547484` ns/iter (`± 923681`) | `0.62` |
| `batch_points_arrow/generate_message_bundles` | `240791` ns/iter (`± 458`) | `336960` ns/iter (`± 1141`) | `0.71` |
| `batch_points_arrow/generate_messages` | `5122` ns/iter (`± 15`) | `6376` ns/iter (`± 28`) | `0.80` |
| `batch_points_arrow/encode_log_msg` | `265927` ns/iter (`± 1247`) | `453146` ns/iter (`± 1968`) | `0.59` |
| `batch_points_arrow/encode_total` | `539116` ns/iter (`± 2351`) | `798510` ns/iter (`± 3665`) | `0.68` |
| `batch_points_arrow/decode_log_msg` | `213761` ns/iter (`± 517`) | `392739` ns/iter (`± 1564`) | `0.54` |
| `batch_points_arrow/decode_message_bundles` | `1900` ns/iter (`± 10`) | `2314` ns/iter (`± 6`) | `0.82` |
| `batch_points_arrow/decode_total` | `220580` ns/iter (`± 861`) | `393182` ns/iter (`± 1465`) | `0.56` |
| `arrow_mono_points/insert` | `1792578633` ns/iter (`± 2165172`) | `7518086366` ns/iter (`± 21814282`) | `0.24` |
| `arrow_mono_points/query` | `1605489` ns/iter (`± 12244`) | `1856154` ns/iter (`± 17882`) | `0.86` |
| `arrow_batch_points/insert` | `780085` ns/iter (`± 1119`) | `3334078` ns/iter (`± 13647`) | `0.23` |
| `arrow_batch_points/query` | `15975` ns/iter (`± 15`) | `17120` ns/iter (`± 49`) | `0.93` |
| `arrow_batch_vecs/insert` | `26293` ns/iter (`± 67`) | `47904` ns/iter (`± 197`) | `0.55` |
| `arrow_batch_vecs/query` | `535506` ns/iter (`± 495`) | `522963` ns/iter (`± 1398`) | `1.02` |
| `tuid/Tuid::random` | `34` ns/iter (`± 0`) | `34` ns/iter (`± 0`) | `1` |


</details>

teh-cmc commented on crates/re_arrow_store/src/store.rs:
```diff
@@ -67,9 +67,27 @@ pub type InsertIdVec = SmallVec<[u64; 4]>;
 pub struct DataTypeRegistry(IntMap<ComponentName, DataType>);
 
+impl DataTypeRegistry {
+    // TODO: why is this here tho
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store.rs:
```diff
@@ -67,9 +67,27 @@ pub type InsertIdVec = SmallVec<[u64; 4]>;
+        let inner: &IntMap<K, V> = &self.0;
+
+        let keys_size_bytes = std::mem::size_of::<K>() * inner.len();
+        let values_size_bytes = std::mem::size_of::<V>() * inner.len();
```
no

teh-cmc commented on crates/re_arrow_store/src/store.rs:
```diff
@@ -87,11 +105,32 @@ impl std::ops::DerefMut for DataTypeRegistry {
+pub struct MetadataRegistry<T: Clone>(BTreeMap<RowId, T>);
+
+impl MetadataRegistry<TimePoint> {
+    // TODO: why is this here tho
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,204 @@ impl std::fmt::Display for GarbageCollectionTarget {
 }
 
-// TODO(#1619): Implement garbage collection.
+// TODO: test with
+// 1. clocks infinite
+// 2. canny infinite
+// 3. objectron but not insanely fast
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,204 @@ impl std::fmt::Display for GarbageCollectionTarget {
+    ///
+    /// ## Semantics
+    ///
+    /// Garbage collection works a row-level basis and is driven by `RowId` order, i.e. the order
```
typo + hard doc links

teh-cmc commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,204 @@ impl std::fmt::Display for GarbageCollectionTarget {
+    ///
+    /// The garbage collector is currently unaware of our latest-at semantics, i.e. it will drop
+    /// old data even if doing so would impact the results of recent queries.
+    /// See <issue>.
```
issue

teh-cmc commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,204 @@ impl std::fmt::Display for GarbageCollectionTarget {
+    /// old data even if doing so would impact the results of recent queries.
+    /// See <issue>.
+    //
+    // TODO(cmc): There shouldn't be any need to return the purged `RowId`s, all secondary
```
issue

teh-cmc commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,204 @@ impl std::fmt::Display for GarbageCollectionTarget {
+    // datastructures should be able to purge themselves based solely off of
+    // `DataStore::oldest_time_per_timeline`.
+    //
+    // TODO(cmc): The GC should be aware of latest-at semantics and make sure they are upheld when
```
issue

teh-cmc commented on crates/re_arrow_store/src/store_read.rs:
```diff
@@ -967,6 +967,9 @@ impl IndexedBucketInner {
 
         *is_sorted = true;
+
+        time_range.min = col_time.first().copied().unwrap_or(i64::MAX).into();
```
a note maybe

teh-cmc commented on crates/re_arrow_store/src/store_sanity.rs:
```diff
@@ -112,11 +112,17 @@ impl IndexedTable {
             let total_size_bytes = self.total_size_bytes();
             let total_size_bytes_uncached = self.total_size_bytes_uncached();
             if total_size_bytes != total_size_bytes_uncached {
+                dbg!((total_size_bytes, total_size_bytes_uncached));
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store_sanity.rs:
```diff
@@ -211,6 +212,7 @@ impl IndexedBucket {
             let size_bytes = inner.read().size_bytes;
             let size_bytes_uncached = inner.write().compute_size_bytes();
             if size_bytes != size_bytes_uncached {
+                dbg!((size_bytes, size_bytes_uncached));
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store_stats.rs:
```diff
@@ -51,6 +119,17 @@ impl DataStoreStats {
 // --- Data store ---
 
 impl DataStore {
+    // TODO
```
???

teh-cmc commented on crates/re_arrow_store/tests/data_store.rs:
```diff
@@ -28,6 +28,7 @@ use re_log_types::{
 
 // --- LatestComponentsAt ---
 
+// TODO: stress test GC in here
```
xxx

teh-cmc commented on crates/re_arrow_store/tests/data_store.rs:
```diff
@@ -380,6 +376,10 @@ fn range() {
     for config in re_arrow_store::test_util::all_configs() {
         let mut store = DataStore::new(InstanceKey::name(), config.clone());
         range_impl(&mut store);
+        // TODO
```
xxx

teh-cmc commented on crates/re_arrow_store/tests/dump.rs:
```diff
@@ -27,6 +29,14 @@ fn data_store_dump() {
 
         data_store_dump_impl(&mut store1, &mut store2, &mut store3);
+
+        // TODO
```
xxx

teh-cmc commented on crates/re_arrow_store/tests/dump.rs:
```diff
@@ -119,6 +131,13 @@ fn data_store_dump_filtered() {
 
         data_store_dump_filtered_impl(&mut store1, &mut store2);
+
+        // TODO
```
xxx

teh-cmc commented on crates/re_data_store/src/log_db.rs:
```diff
@@ -36,7 +36,12 @@ impl Default for EntityDb {
                 InstanceKey::name(),
-                DataStoreConfig::default(),
+                DataStoreConfig {
+                    // NOTE: Empiral testing has shown that 128 is a good balance between sorting
```
typo

teh-cmc commented on crates/re_log_types/src/data_cell.rs:
```diff
@@ -580,8 +650,8 @@ fn data_cell_sizes() {
         // 56 bytes for the inner & outer cells + 3x u64s
-        assert_eq!(80, cell.size_bytes());
-        assert_eq!(80, cell.size_bytes());
+        assert_eq!(144, cell.size_bytes());
+        assert_eq!(144, cell.size_bytes());
```
Schema metadata is not refcounted!! :scream:

teh-cmc commented on crates/re_log_types/src/data_table.rs:
```diff
@@ -118,6 +118,24 @@ impl DataCellColumn {
         }
     }
+
+    // TODO
```
doc

teh-cmc commented on crates/re_log_types/src/data_table.rs:
```diff
@@ -513,6 +531,32 @@ impl DataTable {
         }
     }
+
+    // TODO
```
doc

teh-cmc commented on crates/re_log_types/src/data_table.rs:
```diff
@@ -979,6 +1023,8 @@ impl DataTable {
                 .downcast_ref::<ListArray<i32>>()
                 .ok_or(DataTableError::NotAColumn(component.to_string()))?
                 .iter()
+                // TODO: Pretty sure this clones the schema metadata in each and every array -.-
```
it does

teh-cmc commented on crates/re_log_types/src/time_point/mod.rs:
```diff
@@ -87,6 +87,22 @@ impl TimePoint {
         self
     }
+
+    // TODO
```
doc

teh-cmc commented on examples/rust/objectron/Cargo.toml:
```diff
@@ -8,7 +8,8 @@ publish = false
 
 [dependencies]
-rerun = { workspace = true, features = ["web_viewer"] }
+# rerun = { workspace = true, features = ["web_viewer"] }
```
xxx

teh-cmc commented on crates/re_arrow_store/src/store.rs:
````diff
@@ -449,6 +519,7 @@ impl Default for IndexedBucketInner {
 /// ```
 //
 // TODO(#1524): inline visualization once it's back to a manageable state
+// TODO: timeless should be row-id ordered too then
````
issue

emilk commented on crates/re_arrow_store/tests/correctness.rs:
```diff
@@ -307,64 +310,40 @@ fn gc_correct() {
+    assert_eq!(DataStoreStats::default(), stats_diff);
+
+    if let err @ Err(_) = store.sanity_check() {
+        store.sort_indices_if_needed();
+        eprintln!("{store}");
```
Doesn't `Display` call `sort_indices_if_needed`? It seems to me that a `Display` implementation should be consistent like that

teh-cmc replied:
It's extremely useful to be able to visualize the store exactly as it is when you need to; having the `Display` impl mutate it into something else would defeat that purpose

emilk commented on crates/re_arrow_store/tests/correctness.rs:
```diff
@@ -307,64 +310,40 @@ fn gc_correct() {
+    assert!(row_ids.is_empty());
+    assert_eq!(DataStoreStats::default(), stats_diff);
+
+    if let err @ Err(_) = store.sanity_check() {
+        store.sort_indices_if_needed();
+        eprintln!("{store}");
+        err.unwrap();
+    }
```
If `Display` would call `sort_indices_if_needed` we could simplify this code quite a bit, here and in many places

```suggestion
    store.sanity_check().unwrap_or_else(|err| panic!("{store}\n\n{err}"));
```

emilk commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -17,4 +23,202 @@ impl std::fmt::Display for GarbageCollectionTarget {
+            + stats_before.metadata_registry_size_bytes) as f64;
+
+        let row_ids = match target {
+            GarbageCollectionTarget::DropAtLeastPercentage(p) => {
+                assert!((0.0..=1.0).contains(&p));
```
It makes no sense to only allow dropping 0-1% of the store

```suggestion
            GarbageCollectionTarget::DropAtLeastFraction(p) => {
                assert!((0.0..=1.0).contains(&p));
```

emilk commented on crates/re_arrow_store/tests/data_store.rs:
```diff
@@ -889,29 +897,19 @@ fn gc_impl(store: &mut DataStore) {
-        // for row_id in &row_ids {
-        //     assert!(store.get_msg_metadata(row_id).is_none());
-        // }
+        assert!(
+            (stats.total_size_bytes as f64 * 0.95 / 3.0) as u64 <= stats_diff.total_size_bytes
+                && stats_diff.total_size_bytes
+                    <= (stats.total_size_bytes as f64 * 1.05 / 3.0) as u64
+        );
```
let's make this readable

```suggestion
        let num_bytes_dropped = stats_diff.total_size_bytes as f64;
        assert!(
            (stats.total_size_bytes as f64 * 0.95 / 3.0) <= num_bytes_dropped
                && num_bytes_dropped <= stats.total_size_bytes as f64 * 1.05 / 3.0
        );
```

emilk commented on crates/re_data_store/src/log_db.rs:
```diff
@@ -36,7 +36,12 @@ impl Default for EntityDb {
-                DataStoreConfig::default(),
+                DataStoreConfig {
+                    // NOTE: Empirical testing has shown that 128 is a good balance between sorting
+                    // and binary search costs with the current GC implementation.
```
Then maybe make that the `DataStoreConfig::default`?

emilk replied:
What has the testing be done on btw - huge tensor stuff, or tiny components like `clock`? does it matter?

teh-cmc replied:
GC is all about finding and sorting buckets; the size of the data itself doesn't impact performance.

emilk commented on crates/re_log_types/src/size_bytes_ext.rs:
```diff
@@ -0,0 +1,90 @@
+
+/// A trait to implement `size_bytes` method on types we don't own.
+pub trait SizeBytesExt {
+    /// Returns an accurate-enough approximation of the heap allocated size of `self`.
```
You should add a note here about how implementors should handle the contents of ref-counted data.

emilk replied:
You sometimes include `std::mem::size_of_val(self)` in the implementations, and sometimes you don't. You should document here what is intended.

emilk replied:
You should also document how to handle size vs capacity of e.g. a `Vec`

emilk commented on crates/re_log_types/src/size_bytes_ext.rs:
```diff
@@ -0,0 +1,90 @@
+impl SizeBytesExt for String {
+    #[inline]
+    fn size_bytes(&self) -> u64 {
+        (std::mem::size_of_val(self) + std::mem::size_of_val(self.as_bytes())) as u64
```
Here you include `std::mem::size_of_val(self)`, but `DataType::size_bytes` does not

teh-cmc replied:
it does?

emilk commented on crates/re_log_types/src/size_bytes_ext.rs:
```diff
@@ -0,0 +1,90 @@
+impl SizeBytesExt for String {
+    #[inline]
+    fn size_bytes(&self) -> u64 {
+        (std::mem::size_of_val(self) + std::mem::size_of_val(self.as_bytes())) as u64
```
```suggestion
        (std::mem::size_of_val(self) + self.capacity()) as u64
```

is more accurate

emilk commented on crates/re_log_types/src/time_point/mod.rs:
```diff
@@ -87,6 +88,22 @@ impl TimePoint {
+
+    /// Returns an accurate-enough approximation of the heap allocated size of `self`.
+    #[inline]
+    pub fn size_bytes(&self) -> u64 {
```
If you would `impl SizeBytesExt` instead here and elsewhere, we would get the benefit of only having to document the semantics of `size_bytes` once

emilk reviewed (changes requested):
`gc_drop_at_least_size_bytes` is way to complex to understand and review, especially since there is no comment explaining the gc algorithm. It should be possible to extract the inner loop or two to its own function.

----

A great test to add would be to use a tracking allocator like the one in `crates/re_data_store/examples/memory_usage.rs` to count the number of allocations/deallocations on the current thread, to validate the manual counting in the data store (both during insert and during gc)

teh-cmc commented on crates/re_arrow_store/src/store_write.rs:
```diff
@@ -185,11 +185,30 @@ impl DataStore {
+            std::collections::btree_map::Entry::Vacant(entry) => {
+                entry.insert(timepoint.clone());
+            }
+            // NOTE: When saving and loading data from disk, it's very possible that we try to
+            // insert data for a single `RowId` in multiple calls!
+            std::collections::btree_map::Entry::Occupied(mut entry) => {
+                let entry = entry.get_mut();
+                for (&timeline, &time) in timepoint {
+                    entry.insert(timeline, time);
```
It shouldn't be possible for different columns of the same `RowId` to be inserted in multiple calls (that would be different log calls and so different `RowId`s).

This comment refers to the case where different timelines of the same `RowId` are being inserted in multiple calls, which currently happens anytime buckets are saved to and loaded from disk, since the same `RowId` will likely appear in different buckets (because buckets are per-timeline).
I.e. `assert!(!entry.insert(timeline, time));` should never fail here.

Of course in practice if end users use the low-level APIs to forcibly re-use `RowId`s across log calls then all bets are off but nothing we can do about that besides printing a warning and moving on.

teh-cmc commented:
There are two issues left that I'll tackle first thing tomorrow morning but this should be reviewable as-is.

- Addressed all PR comments above
- Added benchmarks for the GC
- Basically rewrote everything related to size measurements and made `SizeBytes` a first-class citizen everywhere

emilk commented on crates/re_arrow_store/src/store_gc.rs:
```diff
@@ -109,116 +118,163 @@ impl DataStore {
+    ///
+    /// Returns how many bytes were actually dropped, or zero if the row wasn't found.
+    fn try_drop_row(&mut self, row_id: RowId, time: i64) -> u64 {
+        self.sort();
```
# ⚠️ Quadratic Behavior Detected!

Given in-order insertion, we always remove the first row of the first bucket. We do this iteratively, and each removal requires a resort. This leads to quadratic time complexity, though bounded by the maximum rows per bucket.

I see no easy way out of this, expect to replace all the `SmallVec`s of `IndexedBucketInner` with something more like `VecDeque` and special-case removing from the front, or do to something similar manually (keep the `SmallVec`s, but add a `initial_skip_rows: u64` member to `IndexedBucketInner`).

Since the problem is bounded by the max number of rows per bucket it is _probably_ fine, but it should at least be documented, especially where we configure the max rows per bucket (`DataStoreConfig::indexed_bucket_num_rows`), where we need to ad a big fat warning ("The GC costs increases quadratically with this number, so do not make it too large!!!"). Perhaps there should also be a run-time warning when we detect large buckets.

It would be useful to create a benchmark that hits this behavior using different values for `indexed_bucket_num_rows`

teh-cmc replied:
Yes, that's what was alluded to by "Focused on correctness, millions of opportunities to improve performance in the future" in the intro.

There are plenty of workload specific optimizations possible, the `sort` itself hasn't been optimized at all, etc.

I'll add an issue.

> It would be useful to create a benchmark that hits this behavior using different values for `indexed_bucket_num_rows`

`datastore/num_rows=1000/num_instances=1000/packed=false/gc` should cover that already.

emilk reviewed (commented):
Look a lot better now, much easier to follow what's going on! Not I could actually notice the quadratic behavior in there 😆

teh-cmc commented:
Missed a file :unamused: fixed

teh-cmc commented on crates/re_log_types/src/data_cell.rs:
```diff
@@ -582,9 +578,9 @@ fn data_cell_sizes() {
-        // 56 bytes for the inner & outer cells + 3x u64s
-        assert_eq!(144, cell.size_bytes());
-        assert_eq!(144, cell.size_bytes());
+        // zero-sized + 3x u64s
+        assert_eq!(136, cell.heap_size_bytes());
+        assert_eq!(136, cell.heap_size_bytes());
```
An empty cell of type `InstanceKey` requires 112 bytes of heap space (see test case just above this one) because of the fixed overhead of the datatype's schema, `ComponentName`, etc.

With 3 `InstanceKey`s in the cell, i.e. 3x`u64` that's 112+24=136 bytes of heap space.

emilk commented on crates/re_arrow_store/src/store.rs:
```diff
@@ -23,8 +23,13 @@ pub struct DataStoreConfig {
+    /// indexed bucket gets out of order (e.g. because of new insertions or a GC pass).
     /// This is a tradeoff: less rows means faster sorts at the cost of more metadata overhead.
+    /// In particular:
+    /// - Query performance scales inversely logarithmically to this number (i.e. it gets better
+    ///   the higher this number gets).
```
How so? We do a BTree-serarch to find the correct bucket, and then a binary search within the bucket, so it is logarithmic either way, no?

teh-cmc replied:
The number of rows within a bucket is generally bounded (ignoring extremely pathological cases), while the number of buckets is can grow indefinitely

emilk replied:
But the query time of a N buckets of size 1 is log(N), and the query time of 1 bucket of size N is also log(N), no?

So the query time should be independent of the the size of the buckets afaict (at least within a constant factor)
